torch/optim/_functional.py

```diff
@@ -357,3 +357,51 @@ def asgd(params: List[Tensor],
             ax.add_(param.sub(ax).mul(mu))
         else:
             ax.copy_(param)
+
+
+def radam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          state_steps: List[int],
+          *,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          weight_decay: float,
+          eps: float):
+    r"""Functional API that performs RAdam algorithm computation.
+
+    See :class:`~torch.optim.Adam` for details.
+    """
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        step = state_steps[i]
+
+        bias_correction1 = 1 - beta1 ** step
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        # maximum length of the approximated SMA
+        rho_inf = 2 / (1 - beta2) - 1
+        # compute the length of the approximated SMA
+        rho_t = rho_inf - 2 * step * (beta2 ** step) / bias_correction2
+
+        if rho_t > 5.:
+            # Compute the variance rectification term and update parameters accordingly
+            rect = math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
```

Comment:
are this integers (e.g. 4, 2) related in to the threshold for rho_t?

Reply:
In practice, we can see that they are not super related. As you can see both author of the paper and Swift implementors use the threshold cut-off to be 5 however, 4 and 2 remained the same.

Speaking of theory, authors made analysis based on assumption rho_t >= 4. Basically, it means that relation is the algorithm is valid for cutoff to be larger or equal than 4. Therefore, whatever cut-off we select (4, 5 or etc) these pair (4,2) will still be the same independently. However, we need to keep in mind, that if we select cut-off to be less than 4 the algorithm/ its analysis is no more valid.